### PR TITLE
Show error when admintable settings key contains invalid chars

### PR DESCRIPTION
**Pull request by asdil12**
Ticket: https://progress.opensuse.org/issues/104631

### CODE REVIEW

File: lib/OpenQA/WebAPI/Controller/API/V1/Table.pm [hunk 1]
```diff
@@ -397,9 +397,17 @@ sub _prepare_settings {
     my @keys;
     if ($hp->{settings}) {
         for my $k (keys %{$hp->{settings}}) {
-            $k = trim $k;
             my $value = trim $hp->{settings}->{$k};
-            $k =~ s/[^\]\[0-9a-zA-Z_\+]//g;
+            $k = trim $k;
+            my $invalid_chars_in_key = $k;
+            $invalid_chars_in_key =~ s/[\]\[0-9a-zA-Z_\+]//g;
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:403]: This line probably deserves a comment as well.

**Review by Martchus — COMMENTED** [review]
I suppose tests should be extended.

**Review by okurz — APPROVED** [review]

File: lib/OpenQA/WebAPI/Controller/API/V1/Table.pm [hunk 2]
```diff
@@ -397,9 +397,16 @@ sub _prepare_settings {
...
-            $k =~ s/[^\]\[0-9a-zA-Z_\+]//g;
+            $k = trim $k;
+            my $invalid_chars_in_key = $k;
+            # Remove any allowed chars so that only invalid chars remain
+            $invalid_chars_in_key =~ s/[\]\[0-9a-zA-Z_\+]//g;
+            $invalid_chars_in_key =~ s/(.)(?=.*\1)//g;    # Remove any chars that occurs more than once
+            if ($invalid_chars_in_key ne '') {
+                my $eick = join(', ', map { '<b>' . xml_escape($_) . '</b>' } split('', $invalid_chars_in_key));
+                return sprintf('Invalid characters %s in settings key <b>%s</b>', $eick, xml_escape($k));
+            }
```
Comment (perlpunk) [inline lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:409]:
```suggestion
            my %invalid;
            $k =~ s/([^\]\[0-9a-zA-Z_\+])/$inv{$1} = 1/eg;
            delete $invalid{' '}; # silently remove spaces
            if (keys %invalid) {
                my $eick = join ', ', map { '<b>' . xml_escape($_) . '</b>' } sort keys %invalid;
                return sprintf('Invalid characters %s in settings key <b>%s</b>', $eick, xml_escape($k));
            }
```

Comment (perlpunk) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:409]: Also, if you want to keep your code, as an alternative to a regex:
```
my $invalid_chars_in_key = $k = tr/[]0-9a-zA-Z_+//dr;
```

Comment (asdil12) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm]: hm - this doesn't work as expected as it modifies `$k`.

Comment (asdil12) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm]: also I'm not sure if we want to `delete $invalid{' '}; # silently remove spaces`

Comment (asdil12) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm]: I updated it accordingly.

File: lib/OpenQA/WebAPI/Controller/API/V1/Table.pm [hunk 3]
```diff
@@ -397,9 +397,15 @@ sub _prepare_settings {
     my @keys;
     if ($hp->{settings}) {
         for my $k (keys %{$hp->{settings}}) {
-            $k = trim $k;
             my $value = trim $hp->{settings}->{$k};
-            $k =~ s/[^\]\[0-9a-zA-Z_\+]//g;
+            $k = trim $k;
+            my $ktmp = $k;
+            my %invalid;
+            $ktmp =~ s/([^\]\[0-9a-zA-Z_\+])/$invalid{$1} = 1; ''/eg;
```
Comment (perlpunk) [inline lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:404]: You could save one line:
```suggestion
            my %invalid;
            my $ktmp = $k =~ s/([^\]\[0-9a-zA-Z_\+])/$invalid{$1} = 1; ''/egr;
```

Comment (asdil12) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:404]: No this doesn't work. When doing this, `$ktmp` and `$k` have both the same value afterwards - I just tested it.

Comment (asdil12) [reply lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:404]: Edit: When using the `/r` flag, it actually works.

File: lib/OpenQA/WebAPI/Controller/API/V1/Table.pm [hunk 4]
```diff
@@ -397,9 +397,14 @@ sub _prepare_settings {
     my @keys;
     if ($hp->{settings}) {
         for my $k (keys %{$hp->{settings}}) {
-            $k = trim $k;
             my $value = trim $hp->{settings}->{$k};
-            $k =~ s/[^\]\[0-9a-zA-Z_\+]//g;
+            $k = trim $k;
+            my %invalid;
+            () = $k =~ s/([^\]\[0-9a-zA-Z_\+])/$invalid{$1} = 1; ''/egr;
```
Comment (perlpunk) [inline lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:403]: Ok, I didn't see that you would not need the returned string.
Clever :) but maybe then just use a match:
```suggestion
            @inv{$foo =~ m/([^\]\[0-9a-zA-Z_\+])/g} = ();
```

**Review by perlpunk — APPROVED** [review]

**Pull request merged**